ARF 90113 Align Form 21a Steps 1 and 2 with V4 Designs

williamphelps13 opened the pull request:
## Are you removing, renaming or moving a folder in this PR?
- [x] No, I'm not changing any folders (skip to Summary and delete the rest of this section)
- [ ] Yes, I'm removing, renaming or moving a folder

Did you change site-wide styles, platform utilities or other infrastructure?
- [x] No
- [ ] Yes, and I used the [proxy-rewrite steps](https://github.com/department-of-veterans-affairs/vets-website/tree/main/src/applications/proxy-rewrite#that-sounds-normal-so-whats-the-proxy-all-about) to test the injected header scenario

## Summary

- Summarize the changes that have been made to the platform:
  - components (4)
    - PersonalInformationIntro.jsx - Reorder to align with page order; Remove bullet on employment status 
    - RoleDescription.jsx - add
    - NotInGoodStanding.jsx - update text in alignment with designs; update goBack and goForward urls
    - MilitaryServiceIntro.jsx - Remove "Active" from "Active service" since could cause confusion
  - options.js - add parenthetical to claimsAgent option; Update branchOptions to add 'NOAAA' and 'USPHS'
  - IntroductionPage.jsx - update to most current va-process-list version by converting <li> to <va-process-list-item> and moving headers to header prop
  - pages (14)
    - 01-personal-information-chapter/index.js - update standingWithBar to lawLicense in alignment with new page title; remove employmentStatus and workAddress in alignment with designs
    - role.js - add RoleDescription
    - lawLicense.js - add page schema page and no longer generate with inline yesNoPageSchema since it has a page title (because of direct OGC request) unlike other yesNoPages
    - notInGoodStanding.js - update depends for change made in https://github.com/department-of-veterans-affairs/vets-website/pull/31329 and with above lawLicense change
    - nameDateOfBirth.js - make middle name a required field and add hint to add NA if do not have one
    - contactInformation.js - add hint to phone input
    - homeAddress.js - update page title to 'Primary home address'
    - primaryMailingAddress.js - remove conditional options logic since work option is no longer dependent on employmentStatus (since it is now in step 3) add descriptions
    - otherAddress.js - update depends for change made in https://github.com/department-of-veterans-affairs/vets-website/pull/31329 and page description in alignment with designs
    - workAddress.js - remove since integrating into employers with checkbox
    - militaryServicePages.js - extensive updates including integrating Month Year Date, add currentlyServing checkbox, make character and explanation of discharge on subsequent pages since expandUnder is not recommended for accessibility reasons
    - employmentStatus.js - move to chapter 3
    - 03-employment-history-chapter/index.js - integrate into chapter 3; will align chapter 3 with designs in (integrating in https://github.com/department-of-veterans-affairs/vets-website/pull/31457)
    - 06-background-information-chapter/index.js - update depends for change made in https://github.com/department-of-veterans-affairs/vets-website/pull/31329 
  - test (2)
    - 21a.cypress.spec.js - remove workAddress state select hook, refactor military-service-experiences since character of discharge is on a different page
    - test-data.json - update based on all the changes previously mentioned (fixing chapter 3 tests in https://github.com/department-of-veterans-affairs/vets-website/pull/31457)
  - datePatterns.jsx - update currentOrPastDateUI to be able to be able to be conditionally required based on formData and list loop index:
```
    dateRange: currentOrPastMonthYearDateRangeUI(
      {
        title: 'Service start date',
        hint: 'For example: January 2000',
      },
      {
        title: 'Service end date',
        hint: 'For example: January 2000',
        hideIf: (formData, index) =>
          formData?.militaryServiceExperiences?.[index]?.currentlyServing,
        required: (formData, index) =>
          !formData?.militaryServiceExperiences?.[index]?.currentlyServing,
      },
    ),
```
- Accredited Representative Facing
- ARF owns all change except datePatterns.jsx change mentioned above

## Related issue(s)

- [Align Form 21a Steps 1 and 2 with V4 Designs #90113](https://app.zenhub.com/workspaces/accredited-representative-facing-team-65453a97a9cc36069a2ad1d6/issues/gh/department-of-veterans-affairs/va.gov-team/90113)
- [[ARF] Digital Form 21a](https://app.zenhub.com/workspaces/accredited-representative-facing-team-65453a97a9cc36069a2ad1d6/issues/gh/department-of-veterans-affairs/va.gov-team/83627)

## Testing done

- Before: A user could complete Steps 1 and 2 (that aligned with V3 designs)
- After: A user can complete Steps 1 and 2 (that align with V4 designs) and when they click submit after completing the form (could either run cypress or comment out other steps to make this easier) the data that POSTs is the data they entered
- Demoed to design and eng to ensure that steps 1 and 2 align with V4 designs

## Screenshots
<img width="1657" alt="Screenshot 2024-08-21 at 9 07 27 PM" src="https://github.com/user-attachments/assets/622ef04d-6333-40b6-817d-c7212efd4572">
<img width="1609" alt="Screenshot 2024-08-21 at 9 12 21 PM" src="https://github.com/user-attachments/assets/b8cd38d6-9451-4e41-8e68-22de1ca6aa59">
<img width="1664" alt="Screenshot 2024-08-21 at 9 12 45 PM" src="https://github.com/user-attachments/assets/bfa91344-6edf-484f-824a-67d7a99bf11c">
<img width="1666" alt="Screenshot 2024-08-21 at 9 13 15 PM" src="https://github.com/user-attachments/assets/ceaaec18-fb97-4512-9776-940042006a76">

### AFTER
<img width="830" alt="Screenshot 2024-08-21 at 9 13 40 PM" src="https://github.com/user-attachments/assets/3833f09b-70ae-4d74-9796-e4890638f825">
<img width="1661" alt="Screenshot 2024-08-21 at 9 14 06 PM" src="https://github.com/user-attachments/assets/4b0e0c4f-b73c-4ebb-919b-07b3361a8c5c">
<img width="1671" alt="Screenshot 2024-08-21 at 9 15 16 PM" src="https://github.com/user-attachments/assets/eb61c7d3-dfc2-4b49-9bd6-e896c7fc16e9">
<img width="1667" alt="Screenshot 2024-08-21 at 9 15 48 PM" src="https://github.com/user-attachments/assets/2ae2b21d-cb61-44cd-9b39-0eb31800c721">
<img width="1666" alt="Screenshot 2024-08-21 at 9 16 29 PM" src="https://github.com/user-attachments/assets/714ce798-648f-4853-8f92-4b662ec54f99">

### BEFORE
<img width="784" alt="Screenshot 2024-08-21 at 9 17 07 PM" src="https://github.com/user-attachments/assets/83db25b6-c4f7-4771-90aa-666b2c81b750">
<img width="1676" alt="Screenshot 2024-08-21 at 9 17 41 PM" src="https://github.com/user-attachments/assets/22f63657-e1ba-403a-b4d2-0fb7fab61f92">
<img width="1667" alt="Screenshot 2024-08-21 at 9 17 57 PM" src="https://github.com/user-attachments/assets/99122922-0eeb-4e6b-b7a2-51fbd8e0725e">
<img width="1672" alt="Screenshot 2024-08-21 at 9 18 46 PM" src="https://github.com/user-attachments/assets/8061fde9-5477-413d-aab7-6a8588af43ea">
<img width="1667" alt="Screenshot 2024-08-21 at 9 18 58 PM" src="https://github.com/user-attachments/assets/7d3dc486-b06d-4b97-b998-30529906584c">
<img width="1669" alt="Screenshot 2024-08-21 at 9 19 48 PM" src="https://github.com/user-attachments/assets/235c4a65-7964-4f31-b32d-5349122b9a2e">
<img width="1669" alt="Screenshot 2024-08-21 at 9 20 05 PM" src="https://github.com/user-attachments/assets/bc6c514d-18d8-4e7c-b313-d0d37599dd3e">

## Acceptance criteria

### Quality Assurance & Testing

- [x] I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [x] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [x] Linting warnings have been addressed
- [x] Documentation has been updated ([link to documentation](#) \*if necessary)
- [x] Screenshot of the developed feature is added
- [x] [Accessibility testing](https://depo-platform-documentation.scrollhelp.site/developer-docs/wcag-2-1-success-criteria-and-foundational-testing) has been performed

williamphelps13 committed:
Align Step 1 with V4 Designs

williamphelps13 committed:
Align Step 2 with V4 Designs

tjheffner commented on src/applications/accreditation/21a/pages/01-personal-information-chapter/nameDateOfBirth.js:
```diff
@@ -6,16 +7,21 @@ import {
+const fullNameMiddleNameHintUI = cloneDeep(fullNameUI());
+fullNameMiddleNameHintUI.middle['ui:hint'] =
+  'If you do not have a middle name, enter “N/A.”';
+fullNameMiddleNameHintUI.middle['ui:required'] = () => true;
```
nifty way to handle this, nice

tjheffner commented on src/applications/accreditation/21a/pages/02-military-service-chapter/militaryServicePages.js:
```diff
@@ -75,44 +77,87 @@ const introPage = {
       {
-        title: 'Active service start date',
+        title: 'Service end date',
+        hideIf: (formData, index) => isCurrentlyServing(formData, index),
```
really like this approach as well. nice!

![Screenshot 2024-08-16 at 11 28 14 AM](https://github.com/user-attachments/assets/69796200-b1aa-4a0d-ba19-ff936e5af4a0)
![Screenshot 2024-08-16 at 11 28 17 AM](https://github.com/user-attachments/assets/728d0969-e23e-4bf1-ab1b-7f1fb4f9230c)

williamphelps13 committed:
Make 'to' not required and update military urls

williamphelps13 committed:
Add date range to subsequent military page titles

williamphelps13 committed:
Replace explanation of discharge page with expandUnder

Reason for reverting is that list loop depends is not working properly (if depends: () => false then skips summary page). Will inquire with forms team.

williamphelps13 committed:
Merge branch 'main' into arf-90113-align-form-21a-steps-1-2-with-v4-designs

williamphelps13 committed:
Add hint to yes/no on summary page to align with designs

williamphelps13 committed:
Update Introduction Page to use latest version Process List

Response to review feedback on previous PR: https://github.com/department-of-veterans-affairs/vets-website/pull/31329

williamphelps13 committed:
Fix option useage

williamphelps13 committed:
Add back claimsAgent label text to align with design

williamphelps13 committed:
Refine isItemIncomplete

williamphelps13 committed:
Update military service intro path and move isCurrentlyServing inline

williamphelps13 committed:
Update e2e testing

williamphelps13 committed:
Fix so that Middle name now accepts hint

williamphelps13 committed:
Update standingWithBar to lawLicense based on addition of page title

williamphelps13 committed:
Minor fixes from demo to design/eng

Update "Home address" to "Primary home address" and two other minor tweaks.

williamphelps13 committed:
Add Explanation of discharge as conditional page

williamphelps13 committed:
Add additional branchOptions to align with designs

williamphelps13 committed:
Align "I am currently..." with designs

williamphelps13 committed:
Remove unneeded military service page headings. Fix isItemIncomplete.

williamphelps13 committed:
Update currentlyServing text

williamphelps13 committed:
Merge branch 'main' into arf-90113-align-form-21a-steps-1-2-with-v4-designs

williamphelps13 committed:
Allow required to be passed into currentOrPastDateUI

williamphelps13 committed:
Integrate MonthYearDate and fix incomplete data on submission bug

williamphelps13 committed:
Update tests for MonthYearDate and to use selectDropdownHook

williamphelps13 committed:
Add missing word in PersonalInformationIntro

williamphelps13 committed:
Update NotInGoodStanding goBack url to /law-license

williamphelps13 committed:
Update middle name implementation for fullName to be inline (more readable?)

williamphelps13 committed:
Merge branch 'main' into arf-90113-align-form-21a-steps-1-2-with-v4-designs

tjheffner commented on src/applications/accreditation/21a/pages/02-military-service-chapter/militaryServicePages.js:
```diff
@@ -13,13 +14,18 @@ import {
 } from '../../constants/options';
+import { formatReviewDate } from '../helpers/formatReviewDate';
+
+const getDateRange = item => {
```
should this move to helpers? Seems like it could be useful for the employment "I still work here" as well

Andrew565 commented:
PR contains failing tests, choosing not to review at this time. - T2 Support

williamphelps13 committed:
Merge branch 'main' into arf-90113-align-form-21a-steps-1-2-with-v4-designs

williamphelps13 committed:
Merge branch 'main' into arf-90113-align-form-21a-steps-1-2-with-v4-designs

williamphelps13 committed:
Merge branch 'main' into arf-90113-align-form-21a-steps-1-2-with-v4-designs

gabezurita reviewed (commented):
LGTM! The CI errors don't seem related to this PR. Perhaps merge `main` and try again?

williamphelps13 committed:
Merge branch 'main' into arf-90113-align-form-21a-steps-1-2-with-v4-designs

williamphelps13 committed:
Merge branch 'main' into arf-90113-align-form-21a-steps-1-2-with-v4-designs

williamphelps13 committed:
Merge branch 'main' into arf-90113-align-form-21a-steps-1-2-with-v4-designs

williamphelps13 committed:
Merge branch 'main' into arf-90113-align-form-21a-steps-1-2-with-v4-designs

williamphelps13 committed:
Merge branch 'main' into arf-90113-align-form-21a-steps-1-2-with-v4-designs

williamphelps13 committed:
Merge branch 'main' into arf-90113-align-form-21a-steps-1-2-with-v4-designs

williamphelps13 committed:
Merge branch 'main' into arf-90113-align-form-21a-steps-1-2-with-v4-designs

williamphelps13 committed:
Revert currentOrPastDateUI ability to pass required since resulting in failing tests in other apps

I will be making this change in its own PR.

williamphelps13 committed:
Merge branch 'arf-90113-align-form-21a-steps-1-2-with-v4-designs' of github.com:department-of-veterans-affairs/vets-website into arf-90113-align-form-21a-steps-1-2-with-v4-designs

gabezurita commented on src/platform/forms-system/src/js/web-component-patterns/datePatterns.jsx:
```diff
@@ -32,7 +32,7 @@ import {
  */
 const currentOrPastDateUI = options => {
-  const { title, errorMessages, required, ...uiOptions } =
+  const { title, errorMessages, ...uiOptions } =
```
Ah, this caused issues with other apps/CI!

gabezurita reviewed (approved):
LGTM!

jamigibbs reviewed (approved):
Thank you for the excellent description! 🚀
